### ISSUE TITLE
[CLOSED] Test-AdfsServerToken fails if AD FS or WAP doesn't support TLS1.0

### ISSUE THREAD

**Issue by madhavpatel6** [Microsoft/adfsToolbox#9]
<a href="https://github.com/maweeras-msft"><img src="https://avatars1.githubusercontent.com/u/16341169?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [maweeras-msft](https://github.com/maweeras-msft)**
_Friday May 11, 2018 at 14:28 GMT_
_Originally opened as https://github.com/Microsoft/adfsManagementTools/issues/8_

----

If an AD FS farm has been hardened for TLS by disabling TLS1.0, TLS 1.1 as per https://docs.microsoft.com/en-us/windows-server/identity/ad-fs/operations/manage-ssl-protocols-in-ad-fs then the Test-AdfsServerToken will fail as invoke-webrequest uses TLS 1.0 by default.

```
PS C:\WINDOWS\system32> $error[0] | fl * -f


PSMessageDetails      :
Exception             : System.Net.WebException: The underlying connection was closed: An unexpected error occurred on
                        a send. ---> System.IO.IOException: Unable to read data from the transport connection: An
                        existing connection was forcibly closed by the remote host. --->
                        System.Net.Sockets.SocketException: An existing connection was forcibly closed by the remote
                        host
                           at System.Net.Sockets.NetworkStream.Read(Byte[] buffer, Int32 offset, Int32 size)
                           --- End of inner exception stack trace ---
                           at System.Net.Sockets.NetworkStream.Read(Byte[] buffer, Int32 offset, Int32 size)
                           at System.Net.FixedSizeReader.ReadPacket(Byte[] buffer, Int32 offset, Int32 count)
                           at System.Net.Security.SslState.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest
                        asyncRequest)
                           at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message,
                        AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.ForceAuthentication(Boolean receiveFirst, Byte[] buffer,
                        AsyncProtocolRequest asyncRequest)
                           at System.Net.Security.SslState.ProcessAuthentication(LazyAsyncResult lazyResult)
                           at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext,
                        ContextCallback callback, Object state, Boolean preserveSyncCtx)
                           at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback
                        callback, Object state, Boolean preserveSyncCtx)
                           at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback
                        callback, Object state)
                           at System.Net.TlsStream.ProcessAuthentication(LazyAsyncResult result)
                           at System.Net.TlsStream.Write(Byte[] buffer, Int32 offset, Int32 size)
                           at System.Net.ConnectStream.WriteHeaders(Boolean async)
                           --- End of inner exception stack trace ---
                           at Microsoft.PowerShell.Commands.WebRequestPSCmdlet.GetResponse(WebRequest request)
                           at Microsoft.PowerShell.Commands.WebRequestPSCmdlet.ProcessRecord()
TargetObject          : System.Net.HttpWebRequest
CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-WebRequest], WebException
FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand
ErrorDetails          :
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at Test-AdfsServerToken, C:\Program
                        Files\WindowsPowerShell\Modules\ADFSDiagnostics\1.1.0\Public\Test-AdfsServerToken.ps1: line 79
                        at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo : {}
```

As a workaround setting TLS1.2 to be used like so then makes this cmdlet work.

[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12

Can we have the logic for the cmdlet updated to handle the error and use TLS 1.2 if required?

**Comment by madhavpatel6** [comment]
<a href="https://github.com/bongiovimatthew-microsoft"><img src="https://avatars3.githubusercontent.com/u/24795311?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [bongiovimatthew-microsoft](https://github.com/bongiovimatthew-microsoft)**
_Monday May 14, 2018 at 15:17 GMT_

----

Thanks for finding this @maweeras-msft,.We'll take a look and see what we can do.

**Comment by madhavpatel6** [comment]
<a href="https://github.com/madhavpatel6"><img src="https://avatars0.githubusercontent.com/u/9576439?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [madhavpatel6](https://github.com/madhavpatel6)**
_Monday May 21, 2018 at 17:12 GMT_

----

Thanks for the feedback, we have addressed this issue in PR #9. You can now specify a switch to indicate which Tls version you want to utilize. `-TestTls10, -TestTls11, or -TestTls12`. If you installed the module via `Install-Module` you can run `Update-Module` and you will have the new changes.
